pennylane/_queuing.py

```diff
@@ -0,0 +1,195 @@
+    classes that expose a queue for Operations (so-called contexts), as well as the
+    interface to said queues. The active contexts contain maximally one QNode and
+    an arbitrary number of other contexts like the :class:`~.OperationRecorder`.
+    """
+
+    # TODO: update docstring
+
+    _active_contexts = []
+    """The list of contexts that are currently active."""
+
+    def __enter__(self):
+        """Adds this instance to the global list of active contexts.
+
+        Returns:
+            QueuingContext: This instance
+        """
+        QueuingContext._active_contexts.append(self)
+
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        """Remove this instance from the global list of active contexts.
+        """
+        QueuingContext._active_contexts.remove(self)
+
+    @abc.abstractmethod
+    def _append(self, obj):
+        """Append an object to this QueuingContext instance.
+
+        Args:
+            obj: The object to be appended
+        """
+
+    @classmethod
+    def append(cls, obj):
+        """Append an object to the queue(s).
+
+        Args:
+            obj: the object to be appended
+        """
+        for context in cls._active_contexts:
+            context._append(obj)  # pylint: disable=protected-access
+
+    @abc.abstractmethod
+    def _remove(self, obj):
+        """Remove an object from this QueuingContext instance.
+
+        Args:
+            obj: the object to be removed
+        """
+
+    @classmethod
+    def remove(cls, obj):
+        """Remove an object from the queue(s) if it is in the queue(s).
+
+        Args:
+            obj: the object to be removed
+        """
+        for context in cls._active_contexts:
+            # We use the duck-typing approach to assume that the underlying remove
+            # behaves like list.remove and throws a ValueError if the operator
+            # is not in the list
+            try:
+                context._remove(obj)  # pylint: disable=protected-access
+            except ValueError:
+                pass
+
+
+class Queue(QueuingContext):
+    """Lightweight class that maintains a basic queue of operations and pre/post-processing steps
+    for representing quantum circuits."""
+
+    def __init__(self):
+        self.queue = []
+
+    def _append(self, obj):
+        self.queue.append(obj)
+
+    def _remove(self, obj):
+        self.queue.remove(obj)
+
+
+class OperationRecorder(Queue):
+    """A template and quantum function inspector,
+    allowing easy introspection of operators that have been
+    applied without requiring a QNode.
+
+    **Example**:
+
+    The OperationRecorder is a context manager. Executing templates
+    or quantum functions stores resulting applied operators in the
+    recorder, which can then be printed.
+
+    >>> weights = qml.init.strong_ent_layers_normal(n_layers=1, n_wires=2)
+    >>>
+    >>> with qml.utils.OperationRecorder() as rec:
```

Comment:
This docstring needs to be updated. We also need to decide how we want the queuing module to be made available to users/developers.

We should either: 

1. import `OperationRecorder` top level in `__init__.py`, or

2. rename `_queuing.py` -> `queuing.py` and add it to the documentation, or

3. Make `queuing/` a package, have separate modules for the operation recorder etc., and add this to the documentation.

I'm not sure (1) is necessary (since `OperationRecorder` and `QueuingContext` has a very specific usecase), and (3) is only needed if we envision the queuing module growing significantly in the future.

Reply:
Currently, [`QueuingContext`](https://pennylane--722.org.readthedocs.build/en/722/code/api/pennylane.QueuingContext.html) is the only object in this file that displays in the docs, because it's the only object import top level.

Reply:
Since this is only the first of several PRs that will modify these files, I'd like to defer all documentatation/organization changes until things are settled in place. Otherwise we run the risk of updating it multiple times before it lands